File: SWEA/D4/1226. ［S／W 문제해결 기본］ 7일차 － 미로1/［S／W 문제해결 기본］ 7일차 － 미로1.py
from collections import deque

def bfs(maze, x, y):
  queue = deque()
  queue.append(((x, y)))

  while queue:
    x, y = queue.popleft()
    visited[x][y] = True

    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if nx < 0 or ny < 0 or nx >= 16 or ny >= 16:
        continue
      if maze[nx][ny] == 1:
        continue
      if maze[nx][ny] == 0 and visited[nx][ny] == False:
        visited[nx][ny] = True
        queue.append((nx, ny))
      
      if maze[nx][ny] == 3:
        return True

  return False    


dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

for tc in range (1, 11):
  t = int(input())
  board = [list(map(int, input())) for _ in range(16)]
  visited = [[False for _ in range(16)] for _ in range(16)]

  for i in range(16):
    for j in range(16):
      if board[i][j] == 2:
        start_x = i
        start_y = j

  if bfs(board, start_x, start_y):
    print(f"#{t} 1")
  else:
    print(f"#{t} 0")

File: SWEA/D4/1226. ［S／W 문제해결 기본］ 7일차 － 미로1/［S／W 문제해결 기본］ 7일차 － 미로1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class Solution {

    private static int[][] maze;
    private static boolean[][] visited;

    private static final int[] dx = {-1, 0, 1, 0};
    private static final int[] dy = {0, 1, 0, -1};

    private static class Pair {
        int x;
        int y;

        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        for (int t = 1; t < 11; t++) {
            int tc = Integer.parseInt(br.readLine());

            maze = new int[16][16];
            visited = new boolean[16][16];

            int sx = 1;
            int sy = 1;

            // 입력 부분
            for (int i = 0; i < 16; i++) {
                String s = br.readLine();

                for (int j = 0; j < 16; j++) {
                    maze[i][j] = s.charAt(j) - '0';
                }
            }

            int res = bfs(sx, sy);

            System.out.println("#" + t + " " + res);

        }

    }

    private static int bfs(int sx, int sy) {
        ArrayDeque<Pair> queue = new ArrayDeque<>();
        queue.addLast(new Pair(sx, sy));

        visited[sx][sy] = true;

        while (!queue.isEmpty()) {
            Pair cur = queue.pollFirst();

            int x = cur.x;
            int y = cur.y;

            if (maze[x][y] == 3) {
                return 1;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx >= 16 || ny >= 16) {
                    continue;
                }

                if (maze[nx][ny] != 1 && !visited[nx][ny]) {
                    queue.addLast(new Pair(nx, ny));
                    visited[nx][ny] = true;

                }
            }
        }

        return 0;
    }
}